File: mecatro_telemetry/Python/src/mecatro_telemetry/send_gains_to_robot.py

GAINS = [1.5,   # MAHONY_TAU
         8.0,   # KP
         0.05,  # KD
         0.05,  # KP_PHI
         0.015, # KD_PHI
         ]

def send_gains_to_robot(gains, serial):
    '''
    Send gains to robot.
        - gains: list of float32 to send
        - serial: serial object
    '''
    send_array = bytearray(4 * len(gains) + 2)
    send_array[0] = 0xFF
    for i, f in enumerate(gains):
        send_array[4 * i +1:4 * (i + 1) + 1] = struct.pack("f", f)

    checksum = sum(send_array[1:-1]) % 256
    send_array[-1] = checksum

    serial.write(send_array)

File: mecatro_telemetry/Matlab/sendGainsToRobot.m
function sendGainsToRobot(gains, serial)
    % Send gains to robot.
    %   - gains: vector of single precision floats to send
    %   - serial: serial object
    
    sendArray = uint8(zeros(1, 4 * length(gains) + 2));
    sendArray(1) = hex2dec('FF');
    
    for i = 1:length(gains)
        floatBytes = typecast(single(gains(i)), 'uint8');
        startIndex = 4 * (i - 1) + 2;
        endIndex = 4 * i + 1;
        sendArray(startIndex:endIndex) = floatBytes;
    end
    
    checksum = mod(sum(sendArray(2:end-1)), 256);
    sendArray(end) = checksum;
    
    fwrite(serial, sendArray);
end
